### DISCUSSION TITLE
How to deal with poor z-resolution?

### DISCUSSION THREAD

**Discussion by orbisvicis** [pyvista/pyvista#2690] [Q&A]
I have a mesh `z = f(x,y)` which over some regions within the required domain is too steep to have any Z-resolution even when using the maximum amount of memory with an [adaptive](https://github.com/python-adaptive/adaptive) algorithm. Think long, skinny triangles stretching up and down in the Z. I need a method to tessellate the mesh I have cleanly and only in the Z. I can't use [pyacvd](https://github.com/pyvista/pyacvd) because `subdivide(3)` introduces only one new Z-subdivision and higher levels time-out; that's not enough for clustering. Marching cubes seems inefficient since I already have a level surface; given the extra search volume I would have to lower the grid and resulting surface resolution. Surface reconstruction from the underlying points won't preserve the surface. I've seen [optimesh's surface mesh smoothing](https://github.com/meshpro/optimesh#surface-mesh-smoothing) which might spread some edge-points into the Z - and I have the gradient function - but it doesn't seem optimal; also I can't install it. If I subdivide any edges over a given Z-length, how do I re-triangulate the surface?

What should I do?

**Comment by darikg** [comment]
maybe [subdivide_adaptive](https://docs.pyvista.org/api/core/_autosummary/pyvista.PolyDataFilters.subdivide_adaptive.html#pyvista.PolyDataFilters.subdivide_adaptive) with an appropriate max_edge_len?

**Reply by orbisvicis** [reply]
Given a regular structured grid, [subdivide_adaptive](https://docs.pyvista.org/api/core/_autosummary/pyvista.PolyDataFilters.subdivide_adaptive.html#pyvista.PolyDataFilters.subdivide_adaptive) works well for flat surfaces but not where those surfaces meet curved geometry. The resulting triangles along the boundaries are very small, very thin, very irregular, and according to [ACVD](https://github.com/valette/ACVD) non-manifold, though that could just be floating-point approximation during the mesh export. Despite this the results are much better than when working with an adaptive triangulation. I'll still need a way to clean up the results:

* method to analyze and fix non-manifold meshes - perhaps [MeshFix](https://github.com/MarcoAttene/MeshFix-V2.1) can handle this.
* method to limit decimation by triangle angle or triangle area
* method to smooth internal points excluding feature and boundary edges - [PolyData.smooth](https://docs.pyvista.org/api/core/_autosummary/pyvista.PolyData.smooth.html#pyvista.PolyData.smooth) can handle this.

In the meantime I've tried a few things. Marching cubes is slow and memory-hungry. I did not expect resampling a surface onto a uniform grid to require as much memory as marching cubes, but apparently it allocates the entire uniform grid into memory even though that's not necessary. Unlike VTK, pyvista doesn't seem to support [Berger-Oliger type AMR datasets](https://docs.paraview.org/en/latest/UsersGuide/understandingData.html#amr-dataset). Building an unstructured grid of equal-area n-gons from a function and its gradient is [harder than it seems](https://softwarerecs.stackexchange.com/questions/83088/equal-area-polygon-mesh-from-gradient-and-curvature), though I wonder if [extract_surface](https://docs.pyvista.org/api/core/_autosummary/pyvista.UnstructuredGrid.extract_surface.html) can help with this. Unlike [ACVD](https://github.com/valette/ACVD), [pyacvd](https://github.com/pyvista/pyacvd) from pyvista does not [guarantee non-manifold geometry](https://github.com/pyvista/pyacvd/issues/15). In fact, pyacvd produces a lot more non-manifold geometry after [subdivide_adaptive](https://docs.pyvista.org/api/core/_autosummary/pyvista.PolyDataFilters.subdivide_adaptive.html#pyvista.PolyDataFilters.subdivide_adaptive). Also ACVD provides demos that can account for surface gradation and perform anisotropic remeshing.

**Comment by orbisvicis** [comment]
I tried this based on your suggestion, which works to a degree but results in a lot of non-manifold geometry due to the irregular input. The main issue is the initial density of the thin triangles. Now that I'm working with PyVista/VTK I'll start from a quad mesh using a structured grid rather than an adaptive triangulation. If that doesn't work, then marching cubes using a uniform grid. If not perhaps a semi-structured grid of rectangles over the surface integral. It's a pity there is no slicing subdivision, to to subdivide on a grid.

```Python
learner = adaptive.Learner2D\
    (sfunc, bounds=[(-800,1200), (-0.1, 0.1)])
learner.load("...")

l_xy, l_z = learner._data_in_bounds()
l_scale = np.ptp(l_xy, axis=0)
l_xy /= l_scale
with warnings.catch_warnings():
    warnings.simplefilter("error")
    qhull = scipy.spatial.Delaunay(l_xy)
#vertices = np.concatenate((qhull.points * l_scale, l_z), axis=1)
vertices = np.concatenate((qhull.points, l_z), axis=1)
count = np.full((qhull.simplices.shape[0], 1), 3)
faces = np.concatenate((count, qhull.simplices), axis=1)
faces = np.hstack(faces)
surf = pv.PolyData(vertices, faces)

s1 = surf.smooth(200)
s2 = s1.decimate_pro(0.8, splitting=False)
s3 = s2.subdivide_adaptive(max_edge_len=0.02)

clus = pyacvd.Clustering(s3)
clus.subdivide(1)
clus.cluster(20000)
clus.plot()
remesh = clus.create_mesh()
remesh.plot(color='w', show_edges=True)
```
